pkg/reconciler/timeout_handler_test.go

```diff
@@ -259,3 +259,108 @@ func TestWithNoFunc(t *testing.T) {
 	testHandler.CheckTimeouts()
 
 }
+
+// TestGetTimeout checks that the timeout calculated for a given taskrun falls
+// back to the default taskrun timeout when none is provided.
+func TestGetTimeout(t *testing.T) {
+	testCases := []struct {
+		description      string
+		inputDuration    *metav1.Duration
+		expectedDuration time.Duration
+	}{
+		{
+			description:      "returns same duration as input when input is not nil",
+			inputDuration:    &metav1.Duration{2 * time.Second},
+			expectedDuration: 2 * time.Second,
+		},
+		{
+			description:      "returns an end time using the default timeout if a TaskRun's timeout is nil",
+			inputDuration:    nil,
+			expectedDuration: defaultTimeout,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			receivedDuration := GetTimeout(tc.inputDuration)
+			if receivedDuration != tc.expectedDuration {
+				t.Errorf("expected %q received %q", tc.expectedDuration.String(), receivedDuration.String())
+			}
+		})
+	}
+}
+
+// TestSetTaskRunTimer checks that the SetTaskRunTimer method correctly calls the TaskRun
+// callback after a set amount of time.
+func TestSetTaskRunTimer(t *testing.T) {
+	taskRun := tb.TaskRun("test-taskrun-arbitrary-timer", testNs, tb.TaskRunSpec(
+		tb.TaskRunTaskRef(simpleTask.Name, tb.TaskRefAPIVersion("a1")),
+		tb.TaskRunTimeout(2*time.Second),
+	), tb.TaskRunStatus(tb.Condition(apis.Condition{
+		Type:   apis.ConditionSucceeded,
+		Status: corev1.ConditionUnknown}),
+		tb.TaskRunStartTime(time.Now().Add(-10*time.Second)),
+	))
+
+	d := test.Data{
+		TaskRuns: []*v1alpha1.TaskRun{taskRun},
+		Tasks:    []*v1alpha1.Task{simpleTask},
+		Namespaces: []*corev1.Namespace{{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testNs,
+			},
+		}},
+	}
+	stopCh := make(chan struct{})
+	c, _ := test.SeedTestData(t, d)
+	observer, _ := observer.New(zap.InfoLevel)
+	testHandler := NewTimeoutHandler(c.Kube, c.Pipeline, stopCh, zap.New(observer).Sugar())
```

Comment:
i just realized something but _this is totally just scope creep_ so I totally understand if you want to ignore me XD

I looked at this test and I was like "why does it need a kube client and a pipeline client"? and after looking into it it looks like those clients only exist in the timeout handler [to support this one call to `CheckTimeouts`](https://github.com/tektoncd/pipeline/blob/11d1ca8cb7a019dda75ed7deea4764d5cdc68aaa/cmd/controller/main.go#L144) - which means that instead of passing the clients into the timeout handler, the `CheckTimeouts` function could take them as arguments :D which if I'm right would simplify the setup required for this test

anyway feel free to ignore this  XD